### PR TITLE
Support source on Variable, allow more global, class level options

### LINKED ISSUE #15: os.environb suppport?
While trying to sort out storing/retrieving binary values in an env var on Heroku, I noticed that ecological only used `os.environ` and not also `os.environb`. 

I don't know if it's useful to add; just wanted to let you know in case you hadn't considered it. It didn't matter in our case; the Heroku CLI didn't support setting the binary value in the first place, so I fell back on using base64 which worked equivalently for both `environ` and `environb`.

### PR DESCRIPTION
**Pull request by marcinzaremba**
Fixes #15

### CODE REVIEW

File: ecological/config.py [hunk 1]
```diff
@@ -198,10 +69,48 @@ def from_metaclass_kwargs(cls, metaclass_kwargs: Dict) -> "Options":
...
+                raise AttributeError(
+                    f"Configuration error: '{self.variable_name}' is not set."
+                ) from e
+            else:
+                return self.default
+
+        try:
+            return self.transform(raw_value, self.wanted_type)
+        except (ValueError, SyntaxError) as e:
+            raise ValueError(f"Invalid configuration for '{self.variable_name}': {e}.")
```
Comment (thilp) [inline ecological/config.py]:
```suggestion
            raise ValueError(f"Invalid configuration for '{self.variable_name}': {e}.") from e
```
makes it clear that this ValueError is a consequence of `e`. I'm not sure whether having `e` in both the ValueError message and the `from` stacktrace is valuable though.

File: ecological/config.py [hunk 2]
```diff
@@ -198,10 +69,48 @@ def from_metaclass_kwargs(cls, metaclass_kwargs: Dict) -> "Options":
...
+        self.wanted_type = wanted_type
+
+    def get(self) -> VariableValue:
+        try:
+            raw_value = self.source[self.variable_name]
+        except KeyError as e:
+            if self.default is _NO_DEFAULT:
+                raise AttributeError(
+                    f"Configuration error: '{self.variable_name}' is not set."
+                ) from e
```
Comment (thilp) [inline ecological/config.py]:
```suggestion
                ) from None
```
I would not bother reporting the KeyError in addition to the more high-level AttributeError.

**Comment by marcinzaremba** [comment]
I tried to factor the behavior out to the `Config` class but felt really bad from OOP point of view to have another class method that just solely relies on attributes from `Variable` and does not belong to the class per se. I also experimented to have this code directly within the main loop but then it felt `load` method is too heavy. It happens  to be a nice compromise between design and practicality; truly pythonic, right?

**Comment by marcinzaremba** [comment]
What's left here:
- [ ] rewrite `Config.load` method documentation
- [ ] describe global `Config` options in `README.rst`

**Comment by marcinzaremba** [comment]
Now waiting for @thilp final review, then apply the fixes, then I think we can release 2.0 if no major issues come up.

File: README.rst [hunk 3]
```diff
@@ -142,10 +81,115 @@ Nested Configuration
...
+- in the case of specyfing options on both levels (variable and class)
+  the variable ones take precedence over class ones,
+- when some options are missing on the variable level their defaults
+  are taken from the class level,
+- it is not necessary to assign an ``ecological.Variable`` instance to
+  change the behavior; it can still be changed on the class level (globally).
+
+Autoloading
+------------
+It is possible to defer/disable autoloading (setting) of variable values by specifying the ``autoload`` option on class definition.
```
Comment (thilp) [inline README.rst]: I would give more context by adding a paragraph before this one. Something like:

> By default, the values of the attributes of your Ecological configuration class are automatically loaded when the class is created.

File: ecological/config.py [hunk 4]
```diff
@@ -198,7 +83,61 @@ def from_metaclass_kwargs(cls, metaclass_kwargs: Dict) -> "Options":
...
+        Fetches a value of variable from the ``self.source`` and invoke the
+        ``self.transform`` operation on it. Falls back to ``self.default``
+        if the value is not found.
+        """
+        try:
+            raw_value = self.source[self.variable_name]
+        except KeyError:
+            if self.default is _NO_DEFAULT:
+                raise AttributeError(
+                    f"Configuration error: '{self.variable_name}' is not set."
```
Comment (thilp) [inline ecological/config.py]:
```suggestion
                    f"Configuration error: {self.variable_name!r} is not set."
```
I recommend using Python's quoting facilities directly.

File: ecological/config.py [hunk 5]
```diff
@@ -198,7 +83,61 @@ def from_metaclass_kwargs(cls, metaclass_kwargs: Dict) -> "Options":
...
+                ) from None
+            else:
+                return self.default
+
+        try:
+            return self.transform(raw_value, self.wanted_type)
+        except (ValueError, SyntaxError) as e:
+            raise ValueError(
+                f"Invalid configuration for '{self.variable_name}'."
```
Comment (thilp) [inline ecological/config.py]:
```suggestion
                f"Invalid configuration for {self.variable_name!r}."
```

File: README.rst [hunk 6]
```diff
@@ -142,10 +81,115 @@ Nested Configuration
...
+Fine-grained Control
+---------------------
+You can control some behavior of how the configuration properties are set.
+
+It can be achieved by providing a ``ecological.Variable`` instance as the default
+value for an attribute or by specifying global options on the class level:
+
+.. code-block:: python
+
+    my_source: Dict = {"KEY1": "VALUE1"}
```
[the remark this answers is not in the document]
Comment (marcinzaremba) [reply README.rst]: Fair enough.

File: README.rst [hunk 7]
```diff
@@ -142,10 +81,115 @@ Nested Configuration
...
+        my_var2: str
+        # ...
+
+All possible options and their meaning can be found in the table below:
+
++-------------------+---------------+-----------------+-------------------------------------------------+-------------------------------------------------------------------+
+| Option            | Class level   | Variable level  | Default                                         | Description                                                       |
++===================+===============+=================+=================================================+===================================================================+
+| ``prefix``        | yes           | no              | ``None``                                        | A prefix that is prepended when a variable name is derived from   |
+|                   |               |                 |                                                 | an attribute name.                                                |
```
[the remark this answers is not in the document]
Comment (marcinzaremba) [reply README.rst]: Good point. Will add it.

File: README.rst (hunk 3 above)
Comment (marcinzaremba) [reply README.rst]: I am not sure about this one. It's already mentioned once in `Getting started` section and that would also repetition here in regard to the next subsection. Also it's in `Advanced` section. WDYT?

Comment (thilp) [reply README.rst]: After a second read, I agree with you — the _Getting started_ section is not far away. I got biased by only reading the diff. :+1:

**Pull request merged**